Add support for datadog distributions metric

ppsreejith opened the pull request:
### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

Thanks for the awesome work you folks do :)

This PR is my attempt to add support for the [distributions](https://docs.datadoghq.com/developers/metrics/types/?tab=distribution) metric (alternate [link](https://docs.datadoghq.com/metrics/distributions/)) to maintain compatibility with Datadog metrics. This would allow us to aggregate metrics on the server rather than on the agent (i.e. aggregate on Influx rather than on Telegraf).

I've added unit tests and tested this on my local Influx and Telegraf setup. Let me know of all the changes required 👍 

Closes #8134

sjwang90 commented:
Awesome! Thanks @ppsreejith. We currently have a backlog of PR reviews for our next 1.16 release which we want out soon. We'll get this reviewed shortly after.

gerilya commented:
> Awesome! Thanks @ppsreejith. We currently have a backlog of PR reviews for our next 1.16 release which we want out soon. We'll get this reviewed shortly after.

@sjwang90 as v1.16 went out approx one month ago, I am wondering whether you have an ETA for merging this nice feature into the master branch. Thank you

srebhan commented on plugins/inputs/statsd/statsd.go:
```diff
@@ -254,6 +262,11 @@ func (s *Statsd) Gather(acc telegraf.Accumulator) error {
+	for _, m := range s.distributions {
+		acc.AddFields(m.name, m.fields, m.tags, now)
+	}
+	s.distributions = []cacheddistributions{}
```
How about
```suggestion
	s.distributions = make([]cacheddistributions, 0)
```

srebhan commented on plugins/inputs/statsd/statsd.go:
```diff
@@ -322,6 +335,7 @@ func (s *Statsd) Start(ac telegraf.Accumulator) error {
 	s.counters = make(map[string]cachedcounter)
 	s.sets = make(map[string]cachedset)
 	s.timings = make(map[string]cachedtimings)
+	s.distributions = []cacheddistributions{}
```
Same here to keep the same style as above
```suggestion
	s.distributions = make([]cacheddistributions, 0)
```

srebhan commented on plugins/inputs/statsd/statsd.go:
```diff
@@ -735,6 +751,15 @@ func parseKeyValue(keyvalue string) (string, string) {
 func (s *Statsd) aggregate(m metric) {
 	switch m.mtype {
+	case "d":
+		fields := make(map[string]interface{})
+		fields[m.field] = m.floatvalue
+		cached := cacheddistributions{
+			name:   m.name,
+			fields: fields,
+			tags:   m.tags,
+		}
```
You are always setting just one field, why do you use a map for this? You could also store the name and value instead of allocating over and over.

ppsreejith replied:
So I updated the `cacheddistributions` interface and replaced field with value. The repeated allocation now happens in the `Gather` method where the accumulator's `addFields` requires a map argument.

srebhan reviewed (changes requested):
Have some comments in the code, nothing big. Please also add the new option to the `sampleConfig` variable!

srebhan commented:
@ppsreejith any news?

ppsreejith commented:
@srebhan Sorry for the (really) long pause (Life's been hectic 😅). I've some time now so I'll pick this up this week 👍

ppsreejith commented:
@srebhan I've addressed the above comments :+1: Let me know if there are more changes.

> Please also add the new option to the sampleConfig variable!

Do you mean the `sampleConfig` defined in the `plugins/inputs/statsd/statsd.go` file? I'm not sure what changes would be required here.

ppsreejith commented:
Also, small side note: I set up a [playground](https://github.com/ppsreejith/telegraf-sandbox) to make it easier to test out Telegraf changes end-to-end (In case this makes it easier for others).

srebhan commented:
@ppsreejith to be honest I can't remember what I meant with 

> Please also add the new option to the sampleConfig variable!

:-) However, as you add new fields it might be good to give the user a chance to enable or disable the new field(s). We want to keep the output series stable between versions by default. Is there a way to enable/disable the _distribution_ fields? If not, could you please add a flag to the config where those fields are disabled by default!?

ppsreejith commented:
@srebhan Got it. I'll add this 👍

ppsreejith commented:
@srebhan I've added a new config flag `datadog_distributions` to enable parsing the distributions metric.

The distributions metric will only be parsed if both the `datadog_extensions` config _and_ the `datadog_distributions` config are enabled. My reasoning is that since the distributions metric type _is_ a datadog extension, we should only enable parsing if the `datadog_extensions` config is enabled. Furthermore, to keep the option of enabling/disabling this feature, I've added the new `datadog_distributions` config.

I've documented this config in the `sampleConfig` variable as well as the Readme file (with links to Datadog's documentation). I've also extended the tests to cover this config enabled parsing. Let me know if this looks good or if there are other changes required :+1:

srebhan reviewed (approved):
Thank you very much for your effort! Looks good to me!
